### ISSUE TITLE
[Discussion] React 

### ISSUE THREAD

**Issue by geekgonecrazy** [RocketChat/Rocket.Chat#1680]
We have interest in using react in Rocket.Chat.  This ticket is to discuss 

* Pros / Cons
* Why you would like to see this
* Suggestions for how to go about this
* Voicing support / Volunteering to contribute

Go forth and discuss :)

**Comment by sampaiodiego** [comment]
Let's bring all @RocketChat/core here =)

As I said, maybe we should wait for MDG define the future of Blaze. If they succesfully implements a blaze-over-react thing, it should be our best path.

**Comment by geekgonecrazy** [comment]
As I see it Blaze 2

Pros:
* Possibly not as big of a learning curve to those that don't know react.
* We can re-use a big portion of our templates. 
Cons:
* Any react pro that comes in, maybe turned off being that its using some weird templating.
* We now have another piece we have to rely on being maintained.  If MDG dumps it we have another mess to clean up.
* Performance might be impacted by going through the blaze stuff.  There is bound to be at least a little.  But we don't know if it will be noticeable or not.

**Comment by engelgabriel** [comment]
Let's wait and see for a month or two. It's hard to make assumptions now. The performance impact may be zero actually, because the template will be compiled/transpiled into react at `build` time. It won't exists in production... it is like CoffeeScript.

**Comment by tvld** [comment]
MDG announced that they will drop Blaze 1 development, so no new issues taken on. Also "Blaze 2" will not materialize. As far as I can tell, they seem to mostly ignore the community concerns on the forum ( @sampaiodiego pointed also to follow: https://forums.meteor.com/t/next-steps-on-blaze-and-the-view-layer/13561  I for one objected to dropping the installed base for the next hot thing as I dont like eco-system splits, but well... seems we have no choice.

It will be React end of story.

**Comment by tvld** [comment]
MDG was just as clear that we will have ES6 instead of Coffee as main common language. So, it might be that this is a possible pivot point we can even taken to our benefit:. What if all we write in ES6 will be React based so we all know that if it is Coffee, it is still Blaze?  Just a thought ;)

**Comment by geekgonecrazy** [comment]
@tvld I've noticed the same thing in the discussions.

I think using something outside of MDG's solutions at this point might actually be to our benefit.  For one we don't have to wait on them to make up their mind. Second, react is moving fast.  I think we might be putting our self at a disadvantage if we limit to what ever version they support.  Kind of how we have to force people to use an old version of node... But that's another subject.

tl;dr; We shouldn't put our eggs in one basket.

**Comment by tvld** [comment]
It's becoming a bit of a nuisance. We like to start contributing to Rocket. But should we learn Coffee? We did not use up till now, so new.  And should we learn Blaze ? Or wait until Rocket adopts React ? 

I think Rocket has no choice and should become opinionated quickly, otherwise it will hunt them as much as it hunts Meteor now...

**Comment by Morgul** [comment]
If Rocket.Chat is going to decide to make a break from Blaze (which at this point, seems like a smart decision), instead of hopping directly onto React, I'd recommend shopping around a little. Weigh the pros/cons of a few frameworks. For one, I've been blown away with Vue lately, and given their 1.0 release, I'd strongly recommend taking a look at it:

* http://vuejs.org/guide/comparison.html#React
* http://blog.evanyou.me/2015/10/25/vuejs-re-introduction/

Just because React has some neat ideas, doesn't mean they're the only kid on the block. As @tvld said:

> ...Rocket has no choice and should become opinionated quickly, otherwise it will hunt them as much as it hunts Meteor now...

And I agree with that 100%. But, be a little picky, and weigh more than a single option.

**Comment by tvld** [comment]
Indeed, also not a big fan of React. I had a look at  and it seems promising. I would also suggest you have a look at Aurelia...  http://www.aurelia.io

**Comment by engelgabriel** [comment]
From Blaze to React video series.

https://www.discovermeteor.com/category/blaze-to-react/

**Comment by tvld** [comment]
@engelgabriel The right way to do it I suppose. Does it mean you decided to go for React full stop? And if we want to learn Meteor to be able to contribute to Rocket, you would suggest us to learn React?

**Comment by engelgabriel** [comment]
Not yet.. we will do 2 other project in ReactJS next month, and after that I'll be able to give my opinion on this :)

**Comment by tvld** [comment]
Ok, we will take that month to learn the structure of Rocket and decide if we should follow your Rocket or go to another one, Aurelia in our case.

Anyway, please do not forget to give direction as soon as you can. People depend on opinions and choices to be able to invest and not doubt or look back all the time ;)

**Comment by engelgabriel** [comment]
I won't forget, and the reason I want to do this other 2 projects first, is exactly because we understand this is a very important decision. We can't afford to be wrong about this.

**Comment by geekgonecrazy** [comment]
@tvld I've actually looked at Aurelia.  I come from doing a lot of Angular 1.x projects, and Aurelia looks very clean.  I need to play around with it a little.

**Comment by tvld** [comment]
@engelgabriel ... how are your ReactJs project going? Any influence expected here?

**Comment by geekgonecrazy** [comment]
@tvld I think our new docs section we're adding is using react. 

I know i've personally been taking a deep dive into react.  I like it.

Also i've been noticing other people using react with meteor a lot more.

It looks like the kadira guys are doing some work with it:
https://voice.kadira.io/meteor-meets-graphql-3cba2e65fd00

**Comment by tvld** [comment]
@geekgonecrazy I suppose we are waiting for an official opinion from RC for this. I still stare in love at Aurelia, but love does not built a future. We need clear lines and stick to it, I think. So, hence the question: what will RC do? ))

**Comment by geekgonecrazy** [comment]
@tvld Yeah I think all of us are still looking.  Kinda like with the es6 discussion.  It really comes down to us having time to use it to form an opinion one way or another.

Aurelia does look very nice.  Reminds me a bit of angular and angular 2

**Comment by tvld** [comment]
@geekgonecrazy "Reminds me a bit of angular and angular 2" Seems logical as Aurelia's founder Ron, was part of Angular 1 core but was not satisfied with Angular 2... so started all over.

**Comment by engelgabriel** [comment]
Here is yet another post to follow to help us make our minds:
https://forums.meteor.com/t/angular-react-and-blaze/16100

**Comment by geekgonecrazy** [comment]
What I read from that is they don't want to put any man power on it any more.  So they are killing the react attempt.

So they encourage react and then they discourage....  That whole post looked like a big wall of text to say one thing.  Canceling their react effort.

I read through a bulk of the comments on that thread.  It looks like many have started down the react path as a result of meteor moving that direction.  Now that its not moving that way they are going to keep moving forward, as they are happy with it.  

It also looked like companies are liking react over blaze, because they can find people with react / nodejs experience a lot easier then you are going to find someone with blaze / meteor experience.

**tl;dr;**
From where I stand... I think we would be better off moving forward with our own then worrying what MDG does.  I think as a community we would be better served choosing a framework that will make the project better.

**Comment by tvld** [comment]
Actually I have been active myself in the discussion, as I never really understood what is going on. I could not believe they would really alienate their complete atmosphere ecosystem, just for the sake of changing. So I just did not understand. I am afraid that by aiming to please everyone, they satisfy no one and get a bloated hard to maintain product. I hope they proof me wrong as I was a believer :)

@geekgonecrazy I believe you are right. Move forward. With the correct steps taken, I believe RocketChat might easily outgrow Meteor in size, cloud and contributions. Better replace one of the building blocks or even the foundation to keep RC reliable, fast and agile :)

**Comment by engelgabriel** [comment]
I believe that RocketChat will outgrow Meteor. We have been developing many workarounds to Meteor's limitations. As we develop more specific code to deal with our edge cases, move to ReactJS, create our own plugin/package system and Meteor 1.3 get more compatible with NPM, I can see we getting into a point when de-meteorising Rocket.Chat may be easier and necessary.

**Comment by engelgabriel** [comment]
4 strategies to survive Meteor in 2016
https://medium.com/@ottovw/4-strategies-to-survive-meteor-in-2016-ac12fe85c678#.gtzhrg436

**Comment by engelgabriel** [comment]
Top 5 Predictions for Meteor in 2016
https://voice.kadira.io/top-5-predictions-for-meteor-in-2016-c7c196eab308#.7qp9i2ecv

**Comment by engelgabriel** [comment]
And.. the last link for reading:
https://github.com/mattkrick/meatier

**Comment by tvld** [comment]
@engelgabriel The Meatier I did not see... but: excellent stuff !  Libraries are the way I suppose, over Frameworks.

**Comment by tonyxiao** [comment]
What would be the migration path? First convert blaze to React and continue to use Meteor with React and over time migrate other pieces?

**Comment by tvld** [comment]
I might be wrong, but Meteor does not seem to even be sure about React. I think it would be best if Rocket would choose it's own pieces and stick to it.

**Comment by engelgabriel** [comment]
I agree with @tvld, we need to think beyond meteor. And that's why moving to React seems to be the right thing to do. We just need to learn more about it ourselves before we start moving.

**Comment by tvld** [comment]
We are in the process of investing and dedicating one or two devs to this full time... Right now a bit in Limbo where to have them trained on. So... looking forward to the Rocket "cook-book" ;)

**Comment by dcworldwide** [comment]
@tvld that is great news. I'm actively watching this thread to see how this develops. I want to embed rocket.chat within my React based product, but we have deferred this task because we have decided to move from Blaze entirely. Where can I learn more about the programme of work? I may also be able to contribute.

**Comment by tvld** [comment]
@dcworldwide I was not implying that the Rocket "cook-book" already exists and if it does, I am not aware of it. But when it is created, I am sure a lot of people could contribute more easily, including ourselves. We for example would probably take "custom theme-ing" on us ;)

**Comment by dcworldwide** [comment]
@tvld Wasn't referring to the cookbook (not sure what that is). Just the React port.

**Comment by codeluggage** [comment]
Sorry to intrude (the issue is open and recent), but I came to this thread searching for a react version of RC. I specifically looked for a react branch I could fork to begin work on a small side project. Is anyone working on anything like that?

And maybe a more general question, is there enough manpower and effort available to support different "view layers" for RC? Being able to give a "--react" (or "--angular" or...) option would be wonderfully easy when actually trying to use RC for something specific.

Slack as a concept is turning into a "lingua franca" of chat interaction, in my opinion. One of my top future uses for RC is integrating it into other projects as the "chat module". It would open for other view layers from other enthusiasts down the line, and let more people fork it for use inside their own projects. Keeping the view layer "pure" is certainly something I look for in my projects, or I just wrap it in react myself.

**Comment by tvld** [comment]
@codeluggage  I also agree that THE trend for 2016 is that chat becoming a platform, enabling all kind of chat modules.

For a platform to become one,  a good platform need to follow standards to the point of being conservative, most important for the view layer to allow integrations. Standards allow other tools to integrate and adapt without other developers forcing to learn something new. 

Blaze never was a standard and React also is not, with jsx as a really weird choice from that perspective. One could argue that if a large corp supports something then it becomes a standard... but.. Facebook never choose to support React and there are signs they are phasing it out even. React was the thing for 2015, hopefully standards will remain for many years

My suggestion would be stick to JavaScript, use something like [Aurelia](http://eisenbergeffect.bluespire.com/introducing-aurelia/) for the view layer, as it is just...well... JavaScript and fully compliant in the browser. Then we can focus on RC become THE chat platform for years to come. We are still in limbo of choose a new chat platform for medical purposes in Europe. We are waiting on RC to come up with strategy on this before we can commit...

Standards .. please... ? ))

**Comment by dcworldwide** [comment]
>  there are signs they are phasing it out

@tvld Hi Tom, I was surprised by this comment. I'm curious if you can link me to what you've read that suggests this?

Personally, I don't believe chat is a platform - for me chat is a feature/component. Selling a platform is harder then selling developers on the idea of integrating with discrete npm modules. Meteor learnt this the hard way. I'd like to see RC move in the later direction. If it were to, it would be easier for projects like my own to adopt parts of RC and for my team to thus contribute back to core.

**Comment by tvld** [comment]
@dcworldwide There are a few issues with FB in this. First they use a often overseen [Patents file](https://github.com/facebook/react/blob/master/PATENTS) saying that if you ever might obtain a patents that is not in FB's interest, you use your license to React. Second, FB never stated that this is long term strategic. They only outsourced it _after_ using it internal for some time and then when technology processed. For example, last year it became clear that ES6 is fully standard in the browser and just as powerful as jsx in many respects, there are many [discussions](http://programmers.stackexchange.com/questions/225400/pros-and-cons-of-facebooks-react-vs-web-components-polymer) ongoing.

 I am not saying React is bad... it just itches a bit

**Comment by engelgabriel** [comment]
Hi @tvld, we do have a strategy for becoming THE chat platform, it is just not well communicated yet.

In short it is based on 5 pillars:

- Openness
  - Keep the MIT licence to allow other organization to build awesome apps on top of our code
  - Drive revenue from the ecosystem, support and hosting rather than _enterprise_ paid features.
- Standards 
  - Moving away from CoffeeScript and embracing **EcmaScript** 
  - Moving away from Atmosphere Packages to **NPM**
  - Implementing **generic DB** connectors to support more storage options.
  - Support **XMPP** and **MATRIX.org** protocols
  - Already supporting **LDAP**, **CAS**, **oAuth**, **SAML**
- Federation 
  - Allows users from different servers to talk to each other just by adding their emails.
- Integration
  - Add UI to allow users create bots to use the internal hubot implementation
  - Create an online Apps/Bots public catalogue
  - Add import/export functionality for all major alternatives 
- Availability
  - Finish our hosting infrastructure already on beta
  - Create deployment packages for all OSes, package management systems and PaaS.

**Comment by tvld** [comment]
@engelgabriel Great! First thanks for taking the time to write it down. And thanks for the excellent strategy that I understand and is clear!

**Comment by dcsan** [comment]
> For a application to become a strong, lasting platform it needs to follow standards 

as an open project (compared to slack) the ability to customize rocketchat *client* with your own widgets is a huge benefit. rather than jumping on react right away, is there any point to evaluate web components or some other packaging that will allow widgets to be more easily ported to RocketChat?

some type of spec around client side embeddable widgets would go a long way to having richer integrations. although this maybe higher level than the templating language, more like a client-side API
[W3C reference from 2013](https://www.w3.org/TR/2013/REC-widgets-apis-20131031/)

**Comment by engelgabriel** [comment]
I agree @dcsan 

We were very excited about Web component, until version 1.0 dropped the best parts of it, IMHO, like the Shadow DOM, because of browser compatibility.

I was so disappointed that I abandoned the idea for a while to deal with my frustrations...

**Comment by dcsan** [comment]
yeh polymer seems to have not lived up to its potential.

on a totally different angle, client side build time is my biggest pain with meteor right now. I think rocketchat could be a killer app for meteor, it's the first real heavy weight app built with it. 
But the [build times are terrible](https://github.com/RocketChat/Rocket.Chat/issues/2255)

There are all kinds of tricks with webpack that might save that, but meteor has chosen not to go that way, perhaps rightly so with the complexity involved. But another thought is to go for a much lighter framework that just doesn't need all the tricks, such as [mithril - 7k gzipped](http://mithril.js.org/comparison.html)

**Comment by engelgabriel** [comment]
Yeah... we are having problems with build time. We have some ideas about this... it will probably get better moving to pure NPM modules, they seem to deal a lot better with _resolveConstraints.

Anyway, also meteor has promised multiple entry points of UI, so rather than just SERVER and CLIENT, you will have SERVER and CLIENT_MOBILE, CLIENT_WEB, CLIENT_ADMIN, etc

This will help a lot to reduce each entry point, and our LiveChat package that at this point is a bit of a hack... it's a separate app, that is compiled together but has its SERVER part deleted.

**Comment by engelgabriel** [comment]
Building plugin `builLivechat` doubles the time...

**Comment by maggo** [comment]
A few days back MDG [launched meteor 1.3](http://info.meteor.com/blog/announcing-meteor-1.3) and with it full featured support for ES2015 imports and npm! That and react being a first party view layer for meteor mean that it's the right time to seriously think about a shift to new frontend technologies.

**Comment by engelgabriel** [comment]
We agree @maggo, it is just a matter of scheduling time to do it.

**Comment by Sing-Li** [comment]
We have accepted ES2015 PRs for a while now.

Are we ready to make a public statement that we will accept UI PRs that are ReactJS based?

**Comment by dcsan** [comment]
have you converted this project to 1.3, fully - ie taking advantage of some of the new NPM module packing?
did it affect the build times?

I've basically dropped doing anything in meteor cos it was too slow builds = painful once a project got to scale. you guys are like a canary in the coalmine - if it's working for you maybe it's usable again.

**Comment by engelgabriel** [comment]
We have not done that yet, we will do in a couple of weeks. We are waiting for others to find more  bugs :)

**Comment by laomouzi** [comment]
Any update on porting RC to meteor 1.3? thanks!

**Comment by jbraun27** [comment]
How about starting a new branch and recruit a crew of people to refactor Coffeescript to ES6?

**Comment by geekgonecrazy** [comment]
@jbraun27 we are moving at a pretty fast rate forward.  So it would be really hard to keep up a second branch.  You might translate an entire package, but while you're doing so someone comes in the coffescript version and adds a bunch of additional functionality.  Then you'd have to backtrack.  This would be very complicated.

I think any es6 push would realistically have to be done by rewriting one package at a time, submitting the pull request and then moving on to the next one.

Also by just rewriting the code as we work on it.  -- This is the strategy we are trying to take.  Any new package, we try writing in es6.  Fixes to existing, would only convert to es6 if we had to do a pretty significant amount of work already in that area.

If you are willing to help out with the effort to convert to es6.. My suggestion is to pick a package to convert and then hit us up on the demo server in the [#contributors](https://demo.rocket.chat/channel/contributors) channel letting us know.  This way we are aware.

**Comment by daniel-centore** [comment]
@engelgabriel With regards to your comment that you are trying to "Allows users from different servers to talk to each other just by adding their emails", how is that going to work? Will you have a central repository of which server each email is associated with?

**Comment by engelgabriel** [comment]
We want to use DNS for the service discovery, like email.

**Comment by daniel-centore** [comment]
What record type would you use for that?

**Comment by engelgabriel** [comment]
Maybe SRV Records, still discussing.. do you have sugestions?

**Comment by daniel-centore** [comment]
No suggestions here; I have no experience with extending DNS. It's unfortunate, though, that this won't work for users who use hosted email services like gmail.

**Comment by engelgabriel** [comment]
Lets include https://vuejs.org/ on this discussion. 

See https://github.com/meteor/blaze/issues/131

**Comment by tvld** [comment]
Disliking Meteor's strategy a year ago, we left for Telegram + Vue. It's an intermediate solution as we eventually need more control and white labeling.  With that experience, i would +1 for "Vlaze": vue for the front end and SSR, Meteor for the infra & Rocketchat as the new "operating system" for the new messaging app paradigm...

**Comment by dcsan** [comment]
why would you use meteor for infra?
how are you integrating telegram with Vue? you wrote your own telegram client or just pushing to their bot api from your own app and asking people to use their client?

+1 for vuejs. lightweight and featured.

**Comment by tvld** [comment]
@dcsan  with "infra" i meant server side and subscriptions, and so on...

We have extension to standard telegram, using their Games features...

**Comment by tvld** [comment]
@dcsan  with "infra" i meant server side and subscriptions, and so on...

We have extension to standard telegram, using their Games features...

**Comment by engelgabriel** [comment]
Some links:
https://github.com/mitar/vue-test
https://github.com/meteor/blaze/issues/131
https://github.com/Akryum/vue-meteor-tracker/issues/3
https://vue-loader.vuejs.org/en/configurations/pre-processors.html

**Comment by gdelavald** [comment]
To add some points in favor of React (and also Vue):
- React components makes the code more modular and reusable, which could work very well with RC packages approach;
- React adoption is very high and should continue like this in the foreseeable future (https://stackoverflow.com/insights/survey/2017#technology as most loved for ~66% and most wanted as ~18%) what can help having more contributions from the community
- Superior performance to most of the view frameworks nowadays (https://www.youtube.com/watch?v=z5e7kWSHWTg shows some comparisons to a Blaze/Angular demo) but of course Vue stands above in this case (https://vuejs.org/v2/guide/comparison.html)
- React.js similarities with React Native could help bring this projects closer (and perhaps migrating from Cordova to Native could be a thing)

Vue is a really strong contender for React, but at this point in time I still think React has the upper hand, with more adoption and the amount of support from the community.
Either way, Vue could be a great improvement as well (https://medium.com/front-end-hacking/why-how-we-migrated-to-vue-from-blaze-c3437b14dbcf)

**Comment by tvld** [comment]
@gdelavald  we feel that Vue already gained the upper hand as the best contender at this time.. but well... we can all have different ideas ;)

**Comment by cyclops24** [comment]
+1 for use React instead of Blaze.

**Comment by Nepoxx** [comment]
There is a lot to consider to make such an important decision.

React has a massive community, vue.js does not.
Vue.js might be more performant than React, but it's not as good as [Inferno](https://github.com/infernojs/inferno), which is very similar to React and almost a drop-in replacement for it (lots of caveats of course, but still). ([benchmarks](http://stefankrause.net/js-frameworks-benchmark5/webdriver-ts/table.html))

You have to factor in "ease of adoption/contribution". React is used everywhere and known by a lot of people, it might not be the best tech but that's not the only factor to consider.

I used Aurelia personally and love it, great technology, and the people behind it are very good, but the adoption rate is abysmal and that comes with its own issues.

Just my two cents.

**Comment by allanlundhansen** [comment]
+1 on react vs vue

* larger ecosystem 
* testability and simpler debugging of components
* functional approach (clean separation of state via redux!)
* larger development team 
* native support
* Better support from MDG

https://medium.com/js-dojo/react-or-vue-which-javascript-ui-library-should-you-be-using-543a383608d

**Comment by janat08** [comment]
I'm against react, unless you got notifications indicator that randomly lights up costing millions. Being open-source software, not having plethora of files for "composabiity" with somewhat made up syntax and redux to boot is helpful.

**Comment by geekgonecrazy** [comment]
i'm a bit confused @janat08 first part of your sentence seems like you are against.. but the end of the second one seems like you are for it :smile:

**Comment by janat08** [comment]
https://github.com/peerlibrary/meteor-blaze-components
Is OOP like react, nowhere as confining, and permits light adoption.

**Comment by mitar** [comment]
I have finished integrating Tracker with Vue. I think the result looks amazing. There are some internals we could continue improving, but I think that interface for developers works well. [See this comment for more information](https://github.com/meteor-vue/vue-meteor-tracker/issues/3#issuecomment-314778618).

**Comment by jgtoriginal** [comment]
After a whole year using Rocket.Chat, we came to the conclusion that it's front end slows us down on development. We needed something faster development wise, but we didn't want to leave behind the back end build in on Rocket.Chat. So we ended up building every new feature into a custom client in React which speaks with the  Rocket.Chat API. This is great, because we keep on using what we had and we enhance it with React. Development wise, we increased our velocity hugely and we are all happy because after the learning curve React makes things easier. So I'd strongly recommend leaving Blaze behind and going for React.

**Comment by tvld** [comment]
@Nepoxx Funny how things can change in a few months: Vue is now well on its way to surpass React in forks and stars (today 74K/81K) in a few months current growth rate. But dont worry... personal preference plays a role as always, and am sure Rocket Chat will balance the options nicely )

**Comment by jgtoriginal** [comment]
I haven't been using Vue, but I think that Front End Development is moving towards that direction, call it Vue, React or even Angular, which is definitely worth going for. 
To me the fact that Rocket.Chat has such a powerful backend and complex frontend, came as a blocker when trying to implement new features fast. So we went for React, which went up and beyond our expectations.
So I think it'd be a huge advantage to leave Blaze behind. And that actually this would translate on the dev community having an extra pro at the time to decide of choosing Rocket.Chat over any other solution.

**Comment by gdelavald** [comment]
Would love to hear more about your experience on this @jgtoriginal

**Comment by jgtoriginal** [comment]
@gdelavald so as you already know, we've been using RocketChat on it's main flavours: web, cordova, iOS, Android. And we quite liked the concept of coding once, but cordova was just too slow, and native wouldn't allow us that. Besides, no matter what we did, every code change would make us wait a bit to rebuild the app. And when it comes to FE this is a lot of time, cause you can't code on the fly as on BE.

So after lots of research, we came to the conclusion that react was our best option for coding things fast and cross platform. But trying to replace RocketChat FE was a bit of an overhead for us. That's why we explored on the API side of things, after all native apps speak to the API and they perform pretty well. So we did a first test with react native, then we did an other micro service in react web, and after the learning curve, we speeded up our velocity by 3 times at least. 

There's a catch though, and it is that we've been using this Views [https://github.com/viewsdx/docs](url) which is an open source templating engine that allows you to build powerful UIs without doing html or css. You just need to write simple instructions for it to build the UI, so the only thing left is to code it's logic on React. 

Long story short, we got now the power of RocketChat API, we still use RocketChat web, and we develop new custom features in React, which makes our life easier. So if RocketChat would go for React, we would actually end up implementing a lot of new features also within the web service.

**Comment by gdelavald** [comment]
@jgtoriginal Awesome to learn this, we are planning our move to renew the FE architecture as well. Our first experience has been developing the React Native app, and it has been great for us.
I'll update this thread once we got more solid news

**Comment by jgtoriginal** [comment]
@gdelavald that's great news really!! I'm eager on that update then. By the way, I think you really should try Views https://github.com/viewsdx/docs out. Actually it's been developed by one of our team members, so I guess we could explore that together. I'm not currently going to much to your demo server so in any case poke me here. Thx very much!

**Comment by kkumaresan** [comment]
Hi, we at Mongrov have developed a react-native mobile client library for Rocket.Chat. Please check this url: https://github.com/mongrov/roverz . Works both on iOS and Android. If you get a chance, do check this project for a comprehensive native mobile app using Rocket.Chat. We welcome your feedback 😃

**Comment by yuanyong** [comment]
@geekgonecrazy Is the react migration on the road map?

**Comment by geekgonecrazy** [comment]
Its on the roadmap I believe.  I don't know about timing.

@ggazzo might be able to speak to projected roadmap or if it still is on roadmap :smile:

**Comment by sandys** [comment]
here's a suggestion - you guys should use https://pagedraw.io/ and https://news.ycombinator.com/item?id=16467387

It generates fully running React code from sketch/photoshop files. It should cut down development time considerably.

**Comment by soeyi** [comment]
And its 2020 now.

I'm new to Blaze, but I have experiences in React and Vue. And I think the development experience is  terrible with Blaze.
I have so much difficulties in understanding where does this props data comes from.

Any plan of migration ?

**Comment by geekgonecrazy** [comment]
The react rewrite is well underway.  Sections such as admin and the setup wizard were first. :)

**Comment by soeyi** [comment]
@geekgonecrazy 
Thanks;
I've seen some code written in React.
Its possible to give an estimated time when the totally  migration can be done  ?

**Comment by geekgonecrazy** [comment]
Keep an eye on the fuselage repo.  It contains the components being developed.  ETA right now i think is it’s done when it’s done :) I’m sure it will move a long quickly now just need a bit of patience :)

A lot of things going on.  The react component set is just one aspect of the exciting things coming.  Keep an eye out for some rich message stuff in the works.  A lot of powerful features in the works
